magento/magento2-page-builder
# magento/magento2-page-builder#419: Nested Links If User Adds Link Attribute & Link Widget to Banner/Slide

### Description

This pull request will fix the issue with nested links in case the user adds a Widget to Banner/Slide element.

Added a new validation to Banner/Slide elements to prohibit the use of Widgets in case of Link attribute is used.


### Fixed Issues

https://github.com/magento/magento2-page-builder/issues/419: [Bug] Nested Links If User Adds Link Attribute & Link Widget to Banner/Slide


### Manual testing scenarios

**Banner Options Window**

1. Create a CMS page
2. Add Banner element to a Row
3. Click on Edit button (cog icon) on Banner Options Panel
4. Set any link type in Content > Link option.
5. Insert any Widget in Content > Message Text
6. Result: should appear message - "Adding widget in content and link in outer element is not allowed."

**Edit Banner Text area**

1. Create a CMS page
2. Add Banner element to a Row
3. Click on Edit button (cog icon) on Options Panel
4. Set any link type in Content > Link option.
5. Click on Save button (top right) in Options Window
6. Insert any Widget in Edit Banner Text area
7. Result: should appear dialog window - "Nested widgets are not allowed"
8. Click on Ok or Close (X) button of dialog window
9. Result: should disappear dialog window and Widget placeholder in Edit Banner Text area


**Slide Options Window**

1. Create a CMS page
2. Add Slider element to a Row
3. Click on Edit button (cog icon) on Slide Options Panel
4. Set any link type in Content > Link option.
5. Insert any Widget in Content > Message Text
6. Result: should appear message - "Adding widget in content and link in outer element is not allowed."

**Edit Slide Text area**

1. Create a CMS page
2. Add Slider element to a Row
3. Click on Edit button (cog icon) on Slide Options Panel
4. Set any link type in Content > Link option.
5. Click on Save button (top right) in Options Window
6. Insert any Widget in Edit Slide Text area
7. Result: should appear dialog window - "Nested widgets are not allowed"
8. Click on Ok or Close (X) button of dialog window
9. Result: should disappear dialog window and Widget placeholder in Edit Slide Text area




### Checklist

---

### Test Plan
#### Manual checks for ticket
* [x] Verify cannot add both content type link attribute and widget to Banner or Slide
  * [x] Check from Stage using inline editor (set link attribute on edit form first)
  * [x] Check from edit form (check when link attribute is added 1st then add widget. Check reverse operation)
  * [x] Verify correct UI and message in new error modal
* [x] Verify user cannot save edit form in case where they have content created from previous release that already has link attribute and widget
  * [x] Setup environment on 2.3.5-p1, create page w/ banner that has add widget & link attribute, change to this branch, upgrade magento, open edit page for page created on 2.3.5-p1, open edit form for banner, verify clicking Save button on edit panel will open error modal with same message that user cannot have link attribute & widget 
* [x] Check that bug doesn't exist for link attribute plus:
  * [x] Each widget type
  * [x] TinyMCE link
  * [x] TinyMCE image link
* [x] Check that user can add multiple TinyMCE links
  * [x] Don't add link attribute but add tinyMCE widgets links, tinyMCE links, tinyMCE image link

#### Automated MFTF Tests
##### Test 1
- Add slider & set link attribute
- On edit form, also add tinyMCE widget, verify error modal
- On edit form, add tinyMCE link, verify error modal
- On edit form, add tinyMCE image, verify error modal
- Close edit form & inline edit slide
- In inline editor, add tinyMCE widget, verify error modal
- In inline editor, add tinyMCE link, verify error modal
- In inline editor, add tinyMCE image, verify error modal
- Open edit form, delete link attribute, add tinyMCE widget, add link attribute, verify error
##### Test 2
- Extend above test for banner

#### Manual Hip Tests
- No new tests needed
- No updated tests needed
